File: frontend/src/lib/types/exceptionAgents.ts
// Types for the autonomous exception-agent surfaces (decision log + stats).
// Mirrors the backend Pydantic shapes in
// `backend/app/schemas/exception_agent.py`. Confidence is a display-only float
// (stored exact as Numeric(5,4) server-side).

import type { MessageKey } from '$lib/i18n/messages';

export interface AgentDecision {
	id: string;
	exception_id: string;
	invoice_id: string;
	exception_type: string;
	action_taken: 'auto_resolved' | 'escalated' | 'no_action';
	confidence: number;
	rationale: string | null;
	changes: Record<string, { old: string; new: string }> | null;
	autonomy_level: string;
	agent_type: string;
	created_at: string;
}

export interface AgentDecisionList {
	items: AgentDecision[];
	total: number;
	page: number;
	page_size: number;
}

export interface AgentStats {
	total_decisions: number;
	auto_resolved: number;
	escalated: number;
	no_action: number;
	resolution_rate: number;
	escalation_rate: number;
	// Accuracy is a placeholder pending a human-overturn signal — `null` until
	// the platform tracks whether an auto-resolution was later reversed.
	accuracy: number | null;
}

/**
 * The i18n key carrying each action's label — never the English string itself.
 *
 * The dashboard renders this badge in its decision log AND as the outcome of a
 * run it just performed, beside a confidence figure and an autonomy level, so a
 * hardcoded map read as a hole in an otherwise-translated panel. The same three
 * labels are the action filter chips, which is why the chips reuse these keys
 * rather than minting their own — a chip that says one thing and the rows it
 * filters another is the drift a shared map removes.
 *
 * Vocabulary owner: `backend/app/services/exception_agents/base.py`'s
 * `ACTION_*` constants. `exceptionAgents.test.ts` reads them.
 */
export const ACTION_LABEL_KEYS: Record<AgentDecision['action_taken'], MessageKey> = {
	auto_resolved: 'exceptions.agents.action.autoResolved',
	escalated: 'exceptions.agents.action.escalated',
	no_action: 'exceptions.agents.action.noAction'
};

/**
 * The message key for an action, or `null` for one this build doesn't know.
 * `agent_decisions.action_taken` is a plain `String(20)`, so a row written by a
 * newer coordinator can carry an action this frontend predates — the caller
 * then renders the raw value rather than an empty badge.
 */
export function agentActionLabelKey(action: string): MessageKey | null {
	return ACTION_LABEL_KEYS[action as AgentDecision['action_taken']] ?? null;
}

/**
 * The org's autonomy setting, lowest authority first —
 * `backend/app/services/exception_agents/autonomy.py::_THRESHOLDS`.
 * `conservative` is "off": its threshold is unreachable, so every exception
 * escalates.
 */
export const AUTONOMY_LEVELS = ['conservative', 'balanced', 'aggressive'] as const;

export type AutonomyLevel = (typeof AUTONOMY_LEVELS)[number];

/**
 * Autonomy-level labels. Rendered in the decision-log row and in the run
 * dialog's facts list; both printed the raw lowercase wire value before.
 */
export const AUTONOMY_LEVEL_LABEL_KEYS: Record<AutonomyLevel, MessageKey> = {
	conservative: 'exceptions.agents.autonomy.conservative',
	balanced: 'exceptions.agents.autonomy.balanced',
	aggressive: 'exceptions.agents.autonomy.aggressive'
};

/**
 * The message key for an autonomy level, or `null` for an unknown one.
 * `resolve_autonomy_level` already falls back to `conservative` when it STAMPS
 * a decision, but the column is a plain `String(20)` and historical rows are
 * read back as written, so the accessor stays tolerant.
 */
export function autonomyLevelLabelKey(level: string): MessageKey | null {
	return AUTONOMY_LEVEL_LABEL_KEYS[level as AutonomyLevel] ?? null;
}

/**
 * What `POST /api/exceptions/{id}/agent-resolve` returns — the exception's new
 * status plus the single append-only `AgentDecision` the coordinator recorded.
 *
 * There is always a decision, whatever the outcome. `escalated` is a NORMAL
 * result, not a failure: the coordinator resolves the org's autonomy threshold,
 * and a recommendation whose confidence doesn't clear it is handed to a human
 * instead of applied. `no_action` likewise means the resolver found nothing it
 * could safely change. Rendering either as an error would teach operators that
 * the safe path is the broken one.
 * See `backend/docs/exception-agents.md`.
 */
export interface AgentResolveResult {
	exception: { id: string; status: string };
	decision: AgentDecision;
}

/**
 * One exception an agent can be run on — the `GET /api/exceptions` row, narrowed
 * to what the runner panel shows and to what the endpoint will actually accept.
 *
 * `invoice_id` is nullable on the wire and the backend 422s an invoice-less
 * exception (a Positive Pay `not_on_file` fraud return has no invoice for an
 * agent to act on — human triage only), so the panel filters those out rather
 * than offering a Run that can only fail.
 */
export interface AgentCandidateException {
	id: string;
	invoice_id: string | null;
	invoice_number: string | null;
	vendor_name: string | null;
	exception_type: string;
	type_label: string;
	severity: string;
	status: string;
	created_at: string;
}

/** The statuses `POST .../agent-resolve` accepts; anything else 409s. */
export const AGENT_RUNNABLE_STATUSES = ['open', 'escalated'] as const;

File: backend/app/api/exception_agents.py
"""Autonomous exception-agent endpoints — run an agent, read the decision log."""

import uuid

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import ROLE_ADMIN, ROLE_AP_MANAGER, require_roles
from app.api.pagination import PaginationParams, paginated, pagination_params
from app.models.agent_decision import AgentDecision
from app.models.exception import Exception as APException
from app.models.organization import Organization
from app.models.user import User
from app.schemas.exception_agent import (
    AgentDecisionListResponse,
    AgentResolveResponse,
    AgentStatsResponse,
)
from app.services.exception_agents import ExceptionNotActionable, run_agent
from app.tenant import apply_entity_scope, get_entity_id, get_tenant, get_tenant_db

router = APIRouter(prefix="/exceptions", tags=["exception-agents"])


def _decision_dict(d: AgentDecision) -> dict:
    return {
        "id": str(d.id),
        "exception_id": str(d.exception_id),
        "invoice_id": str(d.invoice_id),
        "exception_type": d.exception_type,
        "action_taken": d.action_taken,
        "confidence": float(d.confidence),  # display only — stored exact
        "rationale": d.rationale,
        "changes": d.changes,
        "autonomy_level": d.autonomy_level,
        "agent_type": d.agent_type,
        "created_at": d.created_at.isoformat() if d.created_at else "",
    }


@router.get("/agent-decisions", response_model=AgentDecisionListResponse)
async def list_agent_decisions(
    exception_type: str | None = None,
    action_taken: str | None = None,
    pagination: PaginationParams = Depends(pagination_params),
    db: AsyncSession = Depends(get_tenant_db),
    entity_id: uuid.UUID | None = Depends(get_entity_id),
    user: User = Depends(require_roles(ROLE_ADMIN, ROLE_AP_MANAGER)),
):
    q = select(AgentDecision)
    # TEMP-REVERT-FOR-TEST: q = apply_entity_scope(q, AgentDecision, entity_id)
    if exception_type:
        q = q.where(AgentDecision.exception_type == exception_type)
    if action_taken:
        q = q.where(AgentDecision.action_taken == action_taken)
    total = (await db.execute(select(func.count()).select_from(q.subquery()))).scalar() or 0
    q = (
        q.order_by(AgentDecision.created_at.desc())
        .offset(pagination.offset)
        .limit(pagination.limit)
    )
    rows = (await db.execute(q)).scalars().all()
    return paginated([_decision_dict(d) for d in rows], int(total), pagination)


@router.get("/agent-stats", response_model=AgentStatsResponse)
async def agent_stats(
    db: AsyncSession = Depends(get_tenant_db),
    entity_id: uuid.UUID | None = Depends(get_entity_id),
    user: User = Depends(require_roles(ROLE_ADMIN, ROLE_AP_MANAGER)),
):
    q = select(AgentDecision.action_taken, func.count(AgentDecision.id)).group_by(
        AgentDecision.action_taken
    )
    # TEMP-REVERT-FOR-TEST: q = apply_entity_scope(q, AgentDecision, entity_id)
    rows = (await db.execute(q)).all()
    counts = {a: c for a, c in rows}
    total = sum(counts.values())
    auto = counts.get("auto_resolved", 0)
    esc = counts.get("escalated", 0)
    return {
        "total_decisions": total,
        "auto_resolved": auto,
        "escalated": esc,
        "no_action": counts.get("no_action", 0),
        "resolution_rate": round(auto / total, 4) if total else 0.0,
        "escalation_rate": round(esc / total, 4) if total else 0.0,
        # Accuracy needs a human-overturn signal (was an auto-resolution later
        # reversed?). Not tracked in this slice — placeholder, deferred.
        "accuracy": None,
    }


@router.post("/{exception_id}/agent-resolve", response_model=AgentResolveResponse)
async def agent_resolve(
    exception_id: uuid.UUID,
    db: AsyncSession = Depends(get_tenant_db),
    org: Organization = Depends(get_tenant),  # for .settings (autonomy_level)
    user: User = Depends(require_roles(ROLE_ADMIN, ROLE_AP_MANAGER)),
):
    exc = (
        await db.execute(select(APException).where(APException.id == exception_id))
    ).scalar_one_or_none()
    if not exc:
        raise HTTPException(status_code=404, detail="Exception not found")
    if exc.status not in ("open", "escalated"):
        raise HTTPException(status_code=409, detail=f"Cannot run agent from '{exc.status}' status")
    if exc.invoice_id is None:
        # Invoice-less exceptions (e.g. a Positive Pay not_on_file fraud return)
        # have no invoice for an agent to act on — human triage only.
        raise HTTPException(
            status_code=422,
            detail="This exception has no associated invoice and can't be auto-resolved by an agent.",  # noqa: E501
        )

    try:
        result = await run_agent(
            db,
            exception=exc,
            actor_id=user.id,
            org_settings=org.settings or {},
            # The triggering user's REAL roles — so a CFO-gated invoice resolved
            # by a CFO isn't blocked by a hardcoded ap_manager set, and the audit
            # trail's authoriser role matches the actor_id it records.
            actor_roles={r.name for r in user.roles},
        )
    except ExceptionNotActionable as e:
        # Lost a race with a concurrent agent-resolve — the row lock found the
        # exception already moved on. Same 409 the pre-lock check returns.
        raise HTTPException(
            status_code=409, detail=f"Cannot run agent from '{e.status}' status"
        ) from e
    return {
        "exception": {"id": str(exc.id), "status": exc.status},
        "decision": _decision_dict(result.decision),
    }
